### PR TITLE
NIFI-7417: GetAzureCosmosDBRecord processor

### PR DESCRIPTION
**Pull request by sjyang18**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

Functionally it is equivalent to GetMongoRecord Processor in nifi-mongodb-bundle, but this processor will use the Azure Cosmos native sql API.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [X] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [X] Have you written or updated unit tests to verify your changes?
- [X] Have you verified that the full build is successful on JDK 8?
- [X] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [X] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/GetAzureCosmosDBRecord.java [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+        .required(false)
+        .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+        .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+        .defaultValue("10")
+        .build();
+
+    private final static Set<Relationship> relationships;
+    private final static List<PropertyDescriptor> propertyDescriptors;
+    private ComponentLog logger;
+    private final static ObjectMapper mapper;
```
Comment (Lehel44) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/GetAzureCosmosDBRecord.java:108]: Hi, would you please reorder the qualifiers to "public static final" to comply with JSL :) ?

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/GetAzureCosmosDBRecord.java [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+        boolean queryIsSet = context.getProperty(QUERY).isSet();
+        if (!queryIsSet) {
+            final String msg = QUERY.getDisplayName() + " must be set.";
+            result.add(new ValidationResult.Builder().valid(false).explanation(msg).build());
+        }
+        return result;
+    }
+
+    private String getQuery(ProcessContext context, ProcessSession session, FlowFile input) {
```
Comment (Lehel44) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/GetAzureCosmosDBRecord.java:149]: session is never used, would you please remove it?

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/GetAzureCosmosDBRecord.java [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+                    }
+                    count.incrementAndGet();
+                });
+                writer.finishRecordSet();
+                writer.close();
+                out.close();
+                attributes.put("record.count", String.valueOf(count.get()));
+            } catch (SchemaNotFoundException e) {
+                throw new RuntimeException(e);
+            }
```
Comment (Lehel44) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/GetAzureCosmosDBRecord.java:231]: I'd recommend a little refactor here, to break down the complexity of this method. I'd extract the nested try block and the map initialization.

```suggestion
    private void processSession(ProcessSession session, FlowFile input, Map<String, String> attributes, CosmosPagedIterable<JsonNode> response, FlowFile output, String schemaName) throws IOException {
        final Map<String, String> attributeMap = initAttributeMap(input, schemaName);
        try (OutputStream out = session.write(output)) {
            RecordSchema schema = writerFactory.getSchema(attributeMap, null);
            RecordSetWriter writer = writerFactory.createWriter(getLogger(), schema, out, attributeMap);
            final AtomicLong count = new AtomicLong();
            writer.beginRecordSet();

            response.forEach(data -> processData(schema, writer, count, data));
            writer.finishRecordSet();
            writer.close();
            attributes.put("record.count", String.valueOf(count.get()));
        } catch (SchemaNotFoundException e) {
            throw new RuntimeException(e);
        }
    }
    
    private void processData(RecordSchema schema, RecordSetWriter writer, AtomicLong count, JsonNode data) {
        try {
            Map<String,Object> mapObj = mapper.convertValue(data, new TypeReference<Map<String, Object>>(){});
            Record record = new MapRecord(schema, mapObj);
            writer.write(record);
        } catch(IOException | IllegalArgumentException ex) {
            throw new RuntimeException(ex);
        }
        count.incrementAndGet();
    }
```

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/test/java/org/apache/nifi/processors/azure/cosmos/document/ITGetCosmosDocumentRecord.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+import org.junit.Test;
+
+
+public class ITGetCosmosDocumentRecord extends ITAbstractAzureCosmosDBDocument {
+
+    private static final String TEST_COSMOS_QUERY = "select top 100 * from c";
+    private static final String TEST_PARTITION_KEY = "category";
+    private static List<JsonNode> testData;
+    private static int numOfTestData = 10;
+    private static RecordSchema SCHEMA;
```
Comment (Lehel44) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/test/java/org/apache/nifi/processors/azure/cosmos/document/ITGetCosmosDocumentRecord.java:54]: I think these fields can be final, and would you please rename the constants uppercase to comply to Java conventions? E.g.

```suggestion
    private static final List<JsonNode> TEST_DATA;
    private static final int NUM_OF_TEST_DATA = 10;
    private static final RecordSchema SCHEMA;
```

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/test/java/org/apache/nifi/processors/azure/cosmos/document/ITGetCosmosDocumentRecord.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        runner.setProperty(GetAzureCosmosDBRecord.WRITER_FACTORY, "writer");
+    }
+    @Test
+    public void testReadRecords() {
+        runner.setVariable("schema.name", "sample");
+        runner.enqueue(new byte[] {});
+        runner.run();
+
+        List<MockFlowFile> flowFiles = runner.getFlowFilesForRelationship(AbstractAzureCosmosDBProcessor.REL_SUCCESS);
+        assertTrue(flowFiles.size() == 1);
```
Comment (Lehel44) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/test/java/org/apache/nifi/processors/azure/cosmos/document/ITGetCosmosDocumentRecord.java:116]: Optionally, you can possibly use assertEquals here and also in the other test cases.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/GetAzureCosmosDBRecord.java (hunk 1 above)
Comment (jfrazee) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/GetAzureCosmosDBRecord.java:108]: You mean private right?

Comment (Lehel44) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/GetAzureCosmosDBRecord.java:108]: Yes

Comment (sjyang18) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/GetAzureCosmosDBRecord.java]: Thank you for reviewing my PR. I have taken most of suggestions.

Comment (Lehel44) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/GetAzureCosmosDBRecord.java]: Thank you for the improvements!

**Comment by sjyang18** [comment]
@jfrazee Would you take a look at this PR for me?

**Comment by pvillard31** [comment]
@sjyang18 @jfrazee I think this is an interesting additional feature, are you still looking into this?

**Pull request closed without merging**